src/xdp-utils.c

```diff
@@ -572,6 +573,62 @@ cleanup_temp_file (void *p)
   g_free (*pp);
 }
 
+static GKeyFile *
+spawn_validator (const char* const* args,
+                 int                fd)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GKeyFile) key_file = NULL;
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+  int stdin_fd = -1;
+  int stdout_fd = -1;
+  g_autofree char *stdout_path = NULL;
+
+  stdout_fd = memfd_create("spawn-validator-stdout", 0);
+
+  if (stdout_fd == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_warning ("Failed to spawn subprocess: memfd_create: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  stdin_fd = dup (fd);
+  /* Ensure that we read from the beginning of the file */
+  lseek(stdin_fd, 0, SEEK_SET);
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_take_stdout_fd (launcher, stdout_fd);
+  g_subprocess_launcher_take_stdin_fd (launcher, stdin_fd);
+  process = g_subprocess_launcher_spawnv (launcher, args, &error);
+
+  if (!process)
+    {
+      g_warning ("Failed to spawn subprocess: %s", error->message);
+      return NULL;
+    }
+
+  if (!g_subprocess_wait_check (process, NULL, &error))
+    {
+      /* The child's stderr is automatically redirected to stderr, therefore no need to log it here. */
+      g_warning ("Failed to spawn subprocess: %s", error->message);
+      return NULL;
+    }
+
+  stdout_path = g_strdup_printf("/proc/self/fd/%d", stdout_fd);
+
+  key_file = g_key_file_new ();
+  if (!g_key_file_load_from_file (key_file, stdout_path, G_KEY_FILE_NONE, &error))
```

Comment:
The memfd is memory, you don't have to act like it is a file. mmap the memfd fd and pass that to g_key_file_load_from_data.

Reply:
`g_key_file_load_from_bytes()` would probably be even neater, as `GBytes` is the right representation in GLib for an immutable block of data. `g_mapped_file_get_bytes()` is probably also useful to you

Reply:
Neat. Probably makes sense to have a memfd abstraction that returns GBytes then.

Reply:
Yeah that was something we where talking about on matrix. I was thinking something similar to `GMappedFile`.